src/EA.Iws.Domain/ImportNotificationAssessment/Transactions/ImportPaymentTransaction.cs

```diff
@@ -26,21 +27,39 @@ public class ImportPaymentTransaction
         public async Task Save(Guid notificationId, DateTime date, decimal amount, PaymentMethod paymentMethod,
             string receiptNumber, string comments)
         {
-            if (await transactionCalculator.PaymentIsNowFullyReceived(notificationId, amount))
+            var transaction = ImportNotificationTransaction.PaymentRecord(notificationId, date, amount,
+                paymentMethod, receiptNumber, comments);
+            var balance = await transactionCalculator.Balance(transaction.NotificationId)
+               - transaction.Credit.GetValueOrDefault()
+               + transaction.Debit.GetValueOrDefault();
+
+            transactionRepository.Add(transaction);
+
+            if (balance <= 0)
             {
                 var assessment = await assessmentRepository.GetByNotification(notificationId);
+                var transactions = await transactionRepository.GetTransactions(transaction.NotificationId);
+                transactions = transactions.Where(t => t.Credit > 0).OrderByDescending(t => t.Date).ToList();
 
-                if (assessment.Status == ImportNotificationStatus.AwaitingPayment)
+                foreach (var tran in transactions)
                 {
-                    assessment.PaymentComplete(date);
-                }
-                else if (!assessment.Dates.PaymentReceivedDate.HasValue)
-                {
-                    assessment.Dates.PaymentReceivedDate = date;
+                    if (balance == 0)
```

Comment:
Does this actually work when you are reusing the balance variable set on line 32?  Also should the test on balance not be greater than or equal otherwise an over payment will not trigger the update of the date?

Reply:
Yes. So the balance set on line 32 is the "new" balance after adding the new payment. The foreach loop then goes through the entire transactions (including the new payment) in descending order by date.

The check for balance equals to zero is looking for the transaction that has made the balance to be paid in full. So for instance, if balance is ZERO then it means the latest payment, i.e. the first element from the ordered list is the payment received date - hits the break straightaway.

If there's an over payment, then the loop will continue until it hits the zero mark meaning it's found the transaction that has meet the billable amount.

Hope this makes sense?

Reply:
What happens if non of the payments when added together exactly equal the amount owing?

Reply:
It will simply set the payment date to null.